ddl/column.go

```diff
@@ -427,13 +427,118 @@ func (d *ddl) onSetDefaultValue(t *meta.Meta, job *model.Job) error {
 func (d *ddl) onModifyColumn(t *meta.Meta, job *model.Job) error {
 	newCol := &model.ColumnInfo{}
 	oldColName := &model.CIStr{}
-	err := job.DecodeArgs(newCol, oldColName)
+	pos := &ast.ColumnPosition{}
+	err := job.DecodeArgs(newCol, oldColName, pos)
 	if err != nil {
 		job.State = model.JobCancelled
 		return errors.Trace(err)
 	}
 
-	return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	if pos.Tp == ast.ColumnPositionNone {
+		return errors.Trace(d.updateColumn(t, job, newCol, oldColName))
+	}
+
+	tblInfo, err := t.GetTable(job.SchemaID, job.TableID)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	col := findCol(tblInfo.Columns, newCol.Name.L)
+	if col == nil {
+		col = findCol(tblInfo.Columns, oldColName.L)
+	}
+
+	originalState := col.State
+	switch col.State {
+	case model.StatePublic:
+		if err = d.modifyColumnPosition(tblInfo, newCol, oldColName, pos); err != nil {
+			job.State = model.JobCancelled
+			return errors.Trace(err)
+		}
+		// use delete only to prevent add record based on different column order.
+		// otherwise `insert into t value (1, 2)` will add different row and index record.
+		newCol.State = model.StateDeleteOnly
```

Comment:
Well, I think you are considering about `consistency`, while I'm thinking about `ambiguity`.

The ambiguous query will only appear in something like `INSERT t VALUE (1, 2)`. Other insert query don't have such problem, because they have column list. So if schema is `t(a, b)` the result is `a = 1 b = 2`, otherwise result is `a = 2 b = 1`. It is OK to choose one as the result of this query randomly. 

We can change order directly, let some query respect to old schema, while some respect to the new, both of them can be read by subsequent transaction correctly. But if there are some query like the example are dispatched to different node at this time, we cannot tell user why part of them let `a = 1 b = 2` and others make `b = 1 a = 2` stored in database, just after the `ALTER TABLE t MODIFY COLUMN b INT FIRST` have returned successfully.

I added a test which must fail as demo. In this test case, I commit two `INSERT t VALUE (1, 2)`, one commit before ddl job start, while another commit when job's state is delete only. As you can see, we get two different result finally. Both of them is correct, but the final result seems to be ambiguous.

So the key problem is I don't know whether this ambiguity is a big problem in such query, not whether we can result in bad record which can make subsequent transaction get inconsistent value.